src/dataflow/src/render/filter.rs

```diff
@@ -0,0 +1,276 @@
+        &mut self,
+        relation_expr: &MirRelationExpr,
+    ) -> (Collection<G, Row>, Collection<G, DataflowError>) {
+        if let MirRelationExpr::Filter { input, predicates } = relation_expr {
+            // Partition predicates into normal, and temporal lower/upper bounds.
+            let (normal, lower, upper) = extract_temporal(predicates.iter().cloned())
+                .unwrap_or_else(|err| {
+                    panic!("Temporal predicate error: {:?}", err);
+                });
+
+            let (ok_collection, err_collection) = self.collection(input).unwrap();
+
+            let (oks, errs) = ok_collection.inner.flat_map_fallible({
+                let mut datums = DatumVec::new();
+                move |(data, time, diff)| {
+                    let temp_storage = RowArena::new();
+                    let datums_local = datums.borrow_with(&data);
+                    // Ignore any records failing normal predicate evaluation.
+                    let mut passed = true;
+                    for pred in normal.iter() {
+                        passed = passed
+                            && match pred.eval(&datums_local, &temp_storage) {
+                                Err(e) => {
+                                    return Some(Err((DataflowError::from(e), time, diff)))
+                                        .into_iter()
+                                        .chain(None.into_iter());
+                                }
+                                Ok(Datum::True) => true,
+                                _ => false,
+                            }
+                    }
+                    if passed {
+                        // In order to work with times, it is easiest to convert it to an i128.
+                        // This is because our decimal type uses that representation, and going
+                        // from i128 to u64 is even more painful.
+                        let mut time_128 = i128::from(time);
+
+                        // If there is a lower bound, advance `time` to that lower bound.
+                        // This decimal stuff is brittle; let's hope the scale never changes.
+                        for l in lower.iter() {
+                            match l.eval(&datums_local, &temp_storage) {
+                                Err(e) => {
+                                    return Some(Err((DataflowError::from(e), time, diff)))
+                                        .into_iter()
+                                        .chain(None.into_iter());
+                                }
+                                Ok(Datum::Decimal(s)) => {
+                                    if time_128 < s.as_i128() {
+                                        time_128 = s.as_i128();
+                                    }
+                                }
+                                _ => {
+                                    panic!("Non-decimal value in temporal predicate");
```

Comment:
This is triggered by

```
DROP TABLE one_ts CASCADE;
CREATE TABLE one_ts (ts INTEGER);
INSERT INTO one_ts VALUES (EXTRACT('epoch' FROM CAST('2000-01-01' AS TIMESTAMP))*1000);
INSERT INTO one_ts VALUES (EXTRACT('epoch' FROM CAST('2199-12-31' AS TIMESTAMP))*1000);

CREATE MATERIALIZED VIEW one_lower_bound AS SELECT * FROM one_ts WHERE mz_logical_timestamp() >= ts;
SELECT * FROM one_lower_bound;
```

```
2021-02-09T09:07:22.028839Z ERROR panic: timely:work-0: Non-decimal value in temporal predicate
   0: materialized::handle_panic
             at src/materialized/src/bin/materialized/main.rs:569:65
   1: core::ops::function::Fn::call
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/ops/function.rs:70:5
   2: std::panicking::rust_panic_with_hook
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:581:17
   3: std::panicking::begin_panic::{{closure}}
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:506:9
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys_common/backtrace.rs:153:18
   5: std::panicking::begin_panic
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:505:12
   6: dataflow::render::filter::<impl dataflow::render::context::Context<G,expr::relation::MirRelationExpr,repr::row::Row,T>>::render_filter::{{closure}}
             at src/dataflow/src/render/filter.rs:106:37
      <timely::dataflow::stream::Stream<G,D1> as dataflow::operator::StreamExt<G,D1>>::flat_map_fallible::{{closure}}::{{closure}}::{{closure}}::{{closure}}
             at src/dataflow/src/operator.rs:245:62
      core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/ops/function.rs:280:13
   7: core::option::Option<T>::map
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/option.rs:450:29
      <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/iter/adapters/mod.rs:924:9
      <core::iter::adapters::fuse::Fuse<I> as core::iter::adapters::fuse::FuseImpl<I>>::next
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/iter/adapters/fuse.rs:410:9
      <core::iter::adapters::fuse::Fuse<I> as core::iter::traits::iterator::Iterator>::next
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/iter/adapters/fuse.rs:68:9
      <core::iter::adapters::flatten::FlattenCompat<I,U> as core::iter::traits::iterator::Iterator>::next
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/iter/adapters/flatten.rs:269:19
      <core::iter::adapters::flatten::FlatMap<I,U,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/iter/adapters/flatten.rs:52:9
      <timely::dataflow::stream::Stream<G,D1> as dataflow::operator::StreamExt<G,D1>>::flat_map_fallible::{{closure}}::{{closure}}::{{closure}}
             at src/dataflow/src/operator.rs:245:30
      timely::dataflow::operators::generic::handles::InputHandle<T,D,P>::for_each
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/dataflow/operators/generic/handles.rs:92:13
      <timely::dataflow::stream::Stream<G,D1> as dataflow::operator::StreamExt<G,D1>>::flat_map_fallible::{{closure}}::{{closure}}
             at src/dataflow/src/operator.rs:241:17
      <timely::dataflow::stream::Stream<G,D1> as dataflow::operator::StreamExt<G,D1>>::unary_fallible::{{closure}}::{{closure}}
             at src/dataflow/src/operator.rs:200:17
      timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build::{{closure}}::{{closure}}
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/dataflow/operators/generic/builder_rc.rs:123:31
      timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build_reschedule::{{closure}}
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/dataflow/operators/generic/builder_rc.rs:163:26
      <timely::dataflow::operators::generic::builder_raw::OperatorCore<T,L> as timely::scheduling::Schedule>::schedule
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/dataflow/operators/generic/builder_raw.rs:203:9
   8: timely::progress::subgraph::PerOperatorState<T>::schedule
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/progress/subgraph.rs:649:30
   9: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/progress/subgraph.rs:332:26
      <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/progress/subgraph.rs:298:17
  10: timely::progress::subgraph::PerOperatorState<T>::schedule
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/progress/subgraph.rs:649:30
  11: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/progress/subgraph.rs:332:26
      <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/progress/subgraph.rs:298:17
  12: timely::worker::Wrapper::step::{{closure}}
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/worker.rs:691:57
      core::option::Option<T>::map
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/option.rs:450:29
      timely::worker::Wrapper::step
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/worker.rs:691:26
  13: timely::worker::Worker<A>::step_or_park
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/worker.rs:375:38
  14: dataflow::server::Worker<A>::run
             at src/dataflow/src/server.rs:485:13
      dataflow::server::serve::{{closure}}
             at src/dataflow/src/server.rs:246:13
      timely::execute::execute_from::{{closure}}
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/timely/src/execute.rs:383:22
      timely_communication::initialize::initialize_from::{{closure}}
             at /home/pstoev/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/de69cfc/communication/src/initialize.rs:285:33
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys_common/backtrace.rs:137:18
  15: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/thread/mod.rs:464:17
      <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panic.rs:308:9
      std::panicking::try::do_call
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:381:40
      std::panicking::try
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:345:19
      std::panic::catch_unwind
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panic.rs:382:14
      std::thread::Builder::spawn_unchecked::{{closure}}
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/thread/mod.rs:463:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/ops/function.rs:227:5
  16: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/boxed.rs:1042:9
      <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/boxed.rs:1042:9
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys/unix/thread.rs:87:17
  17: start_thread
  18: clone
```

Reply:
Thanks! It looks like `Null` is a possibility here, and I didn't handle that. My understanding is that any Null should fail a comparison, which will drop the record as it does not pass the filter. Is that your expectation too? Looking at the data, you clearly intended to supply some valid timestamps, but I guess we are converting at least one of them to Null for you.

Reply:
Yes, a NULL, if one was explicitly present in the data, should have failed to match .

Reply:
I'm a bit confused, because both of your inputs seem to be non-null, and e.g.
```
materialize=> select * from  (VALUES (EXTRACT('epoch' FROM CAST('2199-12-31' AS TIMESTAMP))*1000)) as INTEGER;
    column1    
---------------
 7258032000000
(1 row)
```
seems to work out fine. I have fixed it so that it drops the record if a null is seen, but at the same time I'm less clear on where the null came from.

Reply:
It is an integer overflow -- multiplying an UNIX epoch time by 1000 causes the integer to overflow . I will open a separate ticket for this, as it is not related to this change.

Reply:
Hooray! In which case a fix is incoming.